R/track_rust_source.R

```diff
@@ -0,0 +1,80 @@
+# See https://github.com/r-lib/pkgbuild/blob/2aaecf838eb023b788b7a31e267ac644d616dae0/R/compile-dll.R
+
+get_library_path <- function(path = ".") {
+  pkg <- desc::desc(rprojroot::find_package_root_file("DESCRIPTION", path = path))
+  pkg_name <- pkg$get("Package")
+  fs::path(
+    glue::glue(
+      "src",
+      "{pkg_name}{.Platform$dynlib.ext}",
+      .sep = .Platform$file.sep
+    )
+  )
+}
+
+get_rust_files <- function(path = ".") {
+  rust_root <- rprojroot::find_package_root_file("src", "rust", path = path)
+  result <- fs::path()
+  cargo_toml_path <- fs::path(rust_root, "Cargo.toml")
+  if (fs::file_exists(cargo_toml_path)) {
+    result <- c(result, cargo_toml_path)
+  }
+  rust_src_subdir <- fs::path(rust_root, "src")
+  if (fs::dir_exists(rust_src_subdir)) {
+    result <- c(
+      result,
+      fs::dir_ls(
+        path = rust_src_subdir,
+        recurse = TRUE,
+        glob = "*rs"
+      )
+    )
+  }
+  result
+}
+
+needs_compilation <- function(path = ".") {
+  library_path <- get_library_path(path)
+
+  # This will likely never happen
```

Comment:
Why not?

Reply:
I originally thought it may hit this. I think that `roxygen2::roxygenize()` invokes pkgbuild, and `pkgbuild::needs_compilation()` checks if the library file is present, and triggers recompilation if it is absent. This happens **before** any roclets are exectued, so even if there is no library before `roxygen2::roxygenize()`, by the time it invokes our roclet, the library has been already built.
The (unlikely) scenario is the library gets removed during `roxygenize()` process. If that happens, I would not trust the output of the pipeline at all.